### PR TITLE
Set of performance improvements

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**

### CODE REVIEW

File: packages/core/src/render3/PERF_NOTES.md [hunk 1]
```diff
@@ -22,6 +22,18 @@ Great reads:
 
  See benchmark [here](https://jsperf.com/mono-vs-megamorphic-property-access).
 
+## Packed vs. holey Array
```
Comment (mhevery) [inline packages/core/src/render3/PERF_NOTES.md]: thanks for adding this!

File: packages/core/src/render3/instructions/shared.ts [hunk 2]
```diff
@@ -644,11 +642,18 @@ export function createTView(
...
-                        .fill(NO_CHANGE, bindingStartIndex) as LView;
+  // We use a special form of array creation ( Array.apply(null, Array(initialViewLength))
+  // to avoid arrays with holes. References:
+  // https://v8.dev/blog/elements-kinds
+  // https://2ality.com/2012/07/apply-tricks.html
+  // https://2ality.com/2013/11/initializing-arrays.html
+  const blueprint = ngDevMode ? new LViewBlueprint !(initialViewLength) :
+                                Array.apply(null, Array(initialViewLength));
+
+  blueprint.fill(null, 0, bindingStartIndex).fill(NO_CHANGE, bindingStartIndex) as LView;
```
Comment (mhevery) [inline packages/core/src/render3/instructions/shared.ts]: I think there is much better way to do this.
```
const blueprint = ngDevMode ? new LViewBlueprint !() : [];
blueprint.fill(null, 0, bindingStartIndex).fill(NO_CHANGE, bindingStartIndex, initialViewLength) as LView;
```

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/core/src/render3/instructions/shared.ts (hunk 2 above)
Comment (pkozlowski-opensource) [reply packages/core/src/render3/instructions/shared.ts]: This won't work as `.fill` doesn't write past the array length. I've pushed an alternative (if we don't want to use the "apply" trick):

```
 const blueprint = ngDevMode ? new LViewBlueprint !() : [];

  for (let i = 0; i < initialViewLength; i++) {
    blueprint.push(i < bindingStartIndex ? null : NO_CHANGE);
  }
```

Comment (mhevery) [reply packages/core/src/render3/instructions/shared.ts]: Really? OK.

**Review by mhevery — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
Adding "cleanup" label to indicate that the size limit needs an update (as per discussion due the team sync), so that `test_aio_local_ivy` CircleCI task passes.

**Comment by pkozlowski-opensource** [comment]
@AndrewKushnir I've updated the size limit, as discussed. I did it in a separate commit and update to the same value as in the @crisbeto PR #31839 so we are not generating conflicts (or can drop the commit in question if needed).

**Review by petebacondarwin — APPROVED** [review]
```
perf(aio): update main bundle size
```

This does not seem like the correct commit message.
- the subject we use is now docs-infra (aio is deprecated as a subject)
- should it be type perf? I would expect build or ci or even test

Also it would be really helpful for future readers if you explain in the commit message why this has been done (and mention the discussion).

**Comment by AndrewKushnir** [comment]
- [VE Presubmit](http://test/OCL:260741330:BASE:260741366:1564506967748:38b9304)
- [Ivy Presubmit](http://test/OCL:260741330:BASE:260741797:1564507067670:27ed8ec3)

**Pull request closed without merging**